<font size="8">Homework 2: AirBnB Document Database</font>

Group number: `1`

Group members:
1. `Afonso Ascensão : 20240684`
2. `André Filipe Silva : 20230972`
3. `Joana Esteves : 20240746`
4. `Susana Reis : 20240567`
5. `Tomás Figueiredo : 20240941`

The Homework 2 is comprised of two parts:
1. Data modelling (15 points).
2. Queries to database to answer the questions (25 points).

<font size="6">1. Data Modelling</font>

<font size="4">Congratulations! You’ve been hired as part of the new Data Engineering and Management team in the AirBNB Business Intelligence department. The company is restructuring due to unsatisfactory performance from the previous teams.

Before leaving, the head of the Data Modelling department highlighted several issues:

**Data Storage**: A lot of data about AirBNB listings is stored in a single document. While this approach has some advantages, it has also caused performance issues. Queries are slow, and the team didn’t apply patterns, which could improve performance by optimizing the data model. Indexes were also not used.

**Reviews Growth**: The number of reviews for AirBNB is growing rapidly. Currently, we overwrite reviews regularly, but the Business Intelligence department will benefit from storing all reviews and analyzing them over time.

**Data Errors**: There are errors in the data collection, such as duplicate data entries and incorrect timestamps for transactions. The new team will need to decide how to fix these issues.

**Your Role**: In your new role, you’ll need to consider how each database query is used, how often it is needed, and its impact on reads and writes. You should update the database schema to optimize for business use cases. Use tools like embedding, linking, indexes, and patterns to improve the data model. You may need to create new fields, documents, or collections. Be sure to document the pattern you’re applying and the reasons behind your decisions, especially when dealing with duplication and risks of outdated data.

**Key tasks include**:

1. Streamlining the data collection process.
2. Cleaning up the data and optimizing what will be returned for each use case.
3. Applying the correct patterns to speed up common queries.
4. Ensuring departments get accurate and relevant information from the database.
5. Sharing the updated data model schema with other departments.

**Good Practices**: [Check Chapter 6, Mastering MongoDB]

1. All newly created fields should have capitalized names.
2. New queries should work with the most up-to-date database version. If you make multiple changes, all queries should still work after the final updates.
3. For some queries, you may need to change the database schema.
4. When you are applying specific patterns, like polymorphic, subset, or bucket, name them accordingly. 
5. Document each major transformation using this format:
*“We applied {transformation name} because {reasoning behind it}. We expect {change/result} based on {observable measure, such as query speed, number of documents returned, index use, etc.}.”*

</font>


## **Joana + Tomás**

**Data Cleanup and Schema Adjustments:** [9 points in total] 

1) Before working on the queries below, review the data and adjust the schema based on the typical use case described.

**Typical Use Case**: The most common use of the database is to show property listing information to customers. A query retrieves a listing document from the database. Currently, retrieving a listing takes too long. Decide what information should be included in a typical query and optimize the structure accordingly. For example, customers usually only need a sample of reviews, not all reviews (even though all reviews are stored). They also don’t need past transaction data. Update the document schema to fit this use case. This might involve creating new collections or documents.

**Data Cleanup**: Review the data for any errors (such as transactions that don’t belong to the listing) or unnecessary duplication, and clean it up where needed.

<font size="6">2. Uses for the database</font>

<font size="4">Different AirBNB departments require different analytics from our common database.
Below are the specific questions from various departments: 

**Standard Difficulty Questions:** [2 points per question]

## **Susana**
2)	Once a month, we reward hosts with recognition. Select three superhosts with at least two listings that can accommodate more than four people.

## **Susana**
3)	The company considers inevsting into property to rent. Which bed type is most common in listings with a waterfront and a dishwasher in New York?

## **Susana**
4)	We're considering hiring someone to write reviews professionally. Who wrote the longest review in New York?

## **Susana**
5)	To assess the security of different areas, what is the biggest and smallest (price-security deposit) difference per number of visitors at a property?

## **André**
6)  Identify areas by whether they are typically used for short breaks, like weekend mini breaks, or whether they are more suitable for long trips. This information support targeted advertising of different customer types. It is not expected to change much over time so we won’t look to update it, we just require current view. What is the average duration of stay (in nights) per type of property per city (you can use the maximum_nights to measure length of stays)? For each property type return the city with the highest and lowest average value.

**Advanced Difficulty Questions (Consider database optimization for these queries):** [3 points per question]

## **Afonso**
7)	We are creating a new webpage for hosts when setting up their account. It will list suggested typical amenities. This data will need to be available every time a host registers a property but is not expected to change very much. The starting point for the list will be all unique amenities currently listed in properties (across all documents). Optimise the database for this use case and show how the data should be queried.

## **Afonso**
8)	We plan to rtack our reviewers better. We want to create a webpage that shows the top 20 reviewers and the count of the number of reviews of each of these reviewers. This webpage should be kept up to date. It should also have a link to return the number of reviews for a given reviewer ID or Name (show how to query for number of reviews by ID or query quickly).

## **André**
9)	For each property we store review scores across different metrics (accuracy, check-in, cleanliness etc). We consider adding more metrics, although there is no clarity on what these will be. We want to be able to easily query the average score across all of these metrics, including any new metrics that might be added without changing the query. Adjust the data model so this can be done and show the query for an example property.

## **André**
10)	We aim to have better access to information about transaction, we wish to develop a search engine that can calculate the average value of transactions in a given period of time quickly for a given property.

## **André**
11)	We wish to have a summary webpage that displays information about our top destinations. This webpage should display for each of the top 10 cities some basic information about our operations in the area (number of properties by type for example, average price by type) but you can choose the metrics. For each of the top 10 cities it should also provide some basic information about the top 3 properties in each city (price, number of review, whatever you think useful) to show an example of the properties available in the area. We would like to keep this webpage up to date as information changes.

**Database updates:** [2 points per question]
## **Joana + Tomás** 
After optimizing the database, show how to complete the following updates. You can create fictional data. Ensure that previous data does not become stale:

12) Add a new property with a new host in one of the top 10 cities. The host selects the top 10 most common amenities to list.

13) Add a new review from one of our top 20 reviewers for this new property.

14) Add a new review metric called 'x_factor' with a score of 10. Show that the average score across all metrics is correctly calculated for this listing, using the previously developed query.


# Setup

1. run Jupyter 
2. run Docker
3. create MongoDB container in Command Prompt: make sure you don't have 'leftover' container from previous runs

        docker run --name mongodb -d -e MONGO_INITDB_ROOT_USERNAME=AzureDiamond -e MONGO_INITDB_ROOT_PASSWORD=hunter2 -p 27017:27017 mongo

4. run Studio3T

5. create New Connection OR reconnect to existing Connection:

    5.A	In Studio3T create Connection:
            a.	Press Connection -> New Connection
            b.	Insert credentials into URI field: mongodb://AzureDiamond:hunter2@localhost:27017
            c.	Press Test Connection
            d.	Assign a name to the Connection (top empty line) -> Save
            e.	Press Connect
            
    5.B In Studio3T reconnect to Connection:
            a. press Connect (top left corner)
            b. choose MongoDB connection  -> press Connect (MongoDB container must be runnning by this time)
            
6.	In Studio3T import the Database:
            a.	Press Import
            b.	Choose BSON mongodump archive 
            c.	Find your database path
            d.	Select All file formats and choose the database file ‘sampledata.archive’
            e.	Press Run: be patient while the database loads
            f.	Inspect the collections and documents

Done!

[PyMongo documentation](https://pymongo.readthedocs.io/en/stable/api/pymongo/collection.html)

In [1]:
# Python Connector

!pip install pymongo
# or #!conda install -y pymongo

from datetime import datetime
from pprint import pprint
import time
from bson.objectid import ObjectId

from pymongo import MongoClient

user="AzureDiamond"
password="hunter2"
host="localhost"
port="27017"
protocol="mongodb"

client = MongoClient(f"{protocol}://{user}:{password}@{host}:{port}")

# Database check
db = client.sample_airbnb 
print(f"Database info: {db}\n")
db.name 

Database info: Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'sample_airbnb')



'sample_airbnb'

# 1. Data Cleanup and Schema Adjustments

## Loading and Inspecting database


In [2]:
# Collections are inside our Database 'sample_analytics'

collection_list = db.list_collection_names()

print(f"The database contains {len(collection_list)} collections")
print(f"All collections: {collection_list[0:]}")
print(f"Collection {collection_list[0]} contains {db[collection_list[0]].count_documents({})} documents")

The database contains 1 collections
All collections: ['listingsAndReviews']
Collection listingsAndReviews contains 5555 documents


In [3]:
# Get one sample
pprint(db.listingsAndReviews.find_one())

{'_id': '10006546',
 'access': 'We are always available to help guests. The house is fully '
           'available to guests. We are always ready to assist guests. when '
           'possible we pick the guests at the airport.  This service transfer '
           'have a cost per person. We will also have service "meal at home" '
           'with a diverse menu and the taste of each. Enjoy the moment!',
 'accommodates': 8,
 'address': {'country': 'Portugal',
             'country_code': 'PT',
             'government_area': 'Cedofeita, Ildefonso, Sé, Miragaia, Nicolau, '
                                'Vitória',
             'location': {'coordinates': [-8.61308, 41.1413],
                          'is_location_exact': False,
                          'type': 'Point'},
             'market': 'Porto',
             'street': 'Porto, Porto, Portugal',
             'suburb': ''},
 'amenities': ['TV',
               'Cable TV',
               'Wifi',
               'Kitchen',
              

In [ ]:
# Get unique fields from the documents 

unique_fields = set()

for document in db.listingsAndReviews.find():
    for field in document.keys():
        unique_fields.add(field)

for field in sorted(unique_fields):
    print(field)

_id
access
accommodates
address
amenities
availability
bathrooms
bed_type
bedrooms
beds
calendar_last_scraped
cancellation_policy
cleaning_fee
description
extra_people
first_review
guests_included
host
house_rules
images
interaction
last_review
last_scraped
listing_url
maximum_nights
minimum_nights
monthly_price
name
neighborhood_overview
notes
number_of_reviews
price
property_type
review_scores
reviews
reviews_per_month
room_type
security_deposit
space
summary
transit
weekly_price


Structure of the document:
- The document is structured in a way where information about images, adress, availability, hosts and reviews is nested inside the listings' documents.
- We will analyze the possibility of creating other collections:
    - Images, adress and availability nested fields hold information that is relevant when a user wants to check a listing and the amount of information is small and unique to each listing. As such, it makes sense to keep them as nested fields to avoid queries across collections.
    - New collections to consider: Hosts and Reviews? 

## Collections

### Hosts:

In [36]:
pipeline = list(
    db.listingsAndReviews.aggregate([
        {"$project": {"host": 1 }},
        {"$limit": 1}
    ]))

pprint(pipeline)

[{'_id': '10006546',
  'host': {'host_about': 'Gostamos de passear, de viajar, de conhecer pessoas '
                         'e locais novos, gostamos de desporto e animais! '
                         'Vivemos na cidade mais linda do mundo!!!',
           'host_has_profile_pic': True,
           'host_id': '51399391',
           'host_identity_verified': True,
           'host_is_superhost': False,
           'host_listings_count': 3,
           'host_location': 'Porto, Porto District, Portugal',
           'host_name': 'Ana&Gonçalo',
           'host_neighbourhood': '',
           'host_picture_url': 'https://a0.muscache.com/im/pictures/fab79f25-2e10-4f0f-9711-663cb69dc7d8.jpg?aki_policy=profile_x_medium',
           'host_response_rate': 100,
           'host_response_time': 'within an hour',
           'host_thumbnail_url': 'https://a0.muscache.com/im/pictures/fab79f25-2e10-4f0f-9711-663cb69dc7d8.jpg?aki_policy=profile_small',
           'host_total_listings_count': 3,
           '

Check nº of hosts, max and average listings per host - to understand degree of duplication:

In [ ]:
# Total nº of hosts 

pipeline = list( 
    db.listingsAndReviews.aggregate([ 
        {"$group": {"_id": "$host.host_id"}}, 
        {"$count": "total_hosts"},
        {"$project": {"_id": False}} 
        ] ) ) 

pprint(pipeline)

[{'total_hosts': 5104}]


In [ ]:
# Total number of listings per host - top 10

pipeline = list( 
    db.listingsAndReviews.aggregate([ 
        {"$group": {"_id": "$host.host_id", "nr_listings": {"$max": "$host.host_total_listings_count"}}}, 
        {"$sort": {"nr_listings": -1}}, 
        {"$limit": 10},
        {"$project": {"_id": False}} 
        ] ) ) 

pprint(pipeline)

[{'nr_listings': 1198},
 {'nr_listings': 799},
 {'nr_listings': 508},
 {'nr_listings': 472},
 {'nr_listings': 429},
 {'nr_listings': 372},
 {'nr_listings': 324},
 {'nr_listings': 320},
 {'nr_listings': 294},
 {'nr_listings': 294}]


In [42]:
# Average number of listings per host

pipeline = list( 
    db.listingsAndReviews.aggregate([ 
        {"$group": {"_id": None, "avg_listings": {"$avg": "$host.host_total_listings_count"}}},
        {"$project": {"_id": False}} 
        ] ) ) 

pprint(pipeline)

[{'avg_listings': 14.405760576057606}]


**Decision**: Keep host related information nested inside the listing document. Points considered and justification:
- **Host related operations vs. listing related operations**: Usually users search by location, price, amenities, so the main logic revolves around listing. The interest for a specific host information comes as secondary from the interest in the listing, meaning that most of the time we will retrieve the listing and need complementary information on the host, rather than retrieve the host and then look at the listings he has. A separate host document would only be beneficial if querying all listings for a specific host is common, since we have more listings related operations, having the host informations inside the listing document ensures no need for additional queries across different collections.

- **Frequency of updating host fields**: Having separate host documents in another collection makes updating fields easier, because we only have to update one document, instead of updating all listing documents where that host is. However, most fields are unlikely to be updated frequently, host information often remains the same,  so keeping them inside the listing document prevents unnecessary complexity. 

- **Duplicated information**: Separate host document prevents duplication, however one listing only has one host and keeping the host information inside the listing does not cause the document to be very long, as such the duplication may be justified considering the previous points. On average each host has 15 listings so the duplication is not very extreme, but we recognize there are some cases where the duplication will be heavy (e.g. max value of 1198).

### Reviews:

In [40]:
pipeline = list(
    db.listingsAndReviews.aggregate([
        {"$project": {"reviews": 1 }},
        {"$limit": 1}
    ]))

pprint(pipeline)

[{'_id': '10006546',
  'reviews': [{'_id': '58663741',
               'comments': 'A casa da Ana e do Gonçalo foram o local escolhido '
                           'para a passagem de ano com um grupo de amigos. '
                           'Fomos super bem recebidos com uma grande simpatia '
                           'e predisposição a ajudar com qualquer coisa que '
                           'fosse necessário.\r\n'
                           'A casa era ainda melhor do que parecia nas fotos, '
                           'totalmente equipada, com mantas, aquecedor e tudo '
                           'o que pudessemos precisar.\r\n'
                           'A localização não podia ser melhor! Não há melhor '
                           'do que acordar de manhã e ao virar da esquina '
                           'estar a ribeira do Porto.',
               'date': datetime.datetime(2016, 1, 3, 5, 0),
               'listing_id': '10006546',
               'reviewer_id': '51483096',
   

Check max and average reviews per listing:

In [41]:
# Total number of reviews per listing - top 10

pipeline = list( 
    db.listingsAndReviews.aggregate([ 
        {"$project": {"listing_id": 1, "reviews_count": {"$size": "$reviews"}}},
        {"$sort": {"reviews_count": -1}}, 
        {"$limit": 10},
        {"$project": {"_id": False}} 
        ] ) ) 

pprint(pipeline)

[{'reviews_count': 533},
 {'reviews_count': 469},
 {'reviews_count': 463},
 {'reviews_count': 420},
 {'reviews_count': 408},
 {'reviews_count': 402},
 {'reviews_count': 399},
 {'reviews_count': 397},
 {'reviews_count': 391},
 {'reviews_count': 385}]


In [43]:
# Average number of reviews per listing

pipeline = list( 
    db.listingsAndReviews.aggregate([ 
         {"$project": {"listing_id": 1, "reviews_count": {"$size": "$reviews"}}},
        {"$group": {"_id": None,"avg_reviews": {"$avg": "$reviews_count"}}},
        {"$project": {"_id": False}} 
        ] ) ) 

pprint(pipeline)

[{'avg_reviews': 26.965256525652567}]


**Decision**: Separate reviews and listings in two collections, so we can have review specific documents. Points considered and justification: 
- **Size of the document**: The listings have on average 27 reviews, with some cases having hundreds of reviews nested in a single listing. The field “comments” often holds values that are long lines of text. Additionally, with time this problem only tends to get worse as new reviews are added. Given this scenario it would be beneficial to have a new collection for review documents.
- **Review related operations vs. listing related operations**: Usually a user will be interested in a specific listing and want to look at reviews related to that listing. This means that a separate reviews document will lead to additional queries across collections in some cases. However, it is also common to have review related queries (e.g. average review score per listing), and if reviews have a separate collection we will only need to load review related information, avoiding loading all the remaining information that is in the listing’s documents.
- **Keep some info in the listing's documents**: Aditionally, we will keep 3 reviews nested inside the listings' documents for efficient querying.
- **Implications**: We need to ensure consistency between collections. For example, when a new review is added we need to update some fields in the listings' documents, like: last_review and number_of_reviews.

## Duplicates

### Listings:

In [ ]:
# Check listings with duplicated ids 

pipeline = list(
    db.listingsAndReviews.aggregate([
        {"$group": { "_id": "$_id", "count": { "$sum": 1 }} },
        { "$match": {"count": { "$gt": 1 }}}
    ])
)

pprint(pipeline)

[]


In [ ]:
# check listings w/ duplicated urls
pipeline = list(
    db.listingsAndReviews.aggregate([
        {"$group": { "_id": "$listing_url", "count": { "$sum": 1 }} },
        { "$match": {"count": { "$gt": 1 }}}
    ])
)

pprint(pipeline)

[]


In [ ]:
# Check listings with duplicates names 

pipeline = list(
    db.listingsAndReviews.aggregate([
        {"$group": { "_id": "$name", "count": { "$sum": 1 }} },
        { "$match": {"count": { "$gt": 1 }}}
    ])
)

pprint(pipeline)

[{'_id': 'Triple room with private bathroom', 'count': 2},
 {'_id': 'Quarto moradia luxo', 'count': 2},
 {'_id': 'Great Location In Wakiki, Walking Distance To Beach, Shopping, '
         'Dining!',
  'count': 2},
 {'_id': 'İstanbul Birden fazla bölümden oluşan bina', 'count': 4},
 {'_id': 'Feel like home', 'count': 2},
 {'_id': 'Elegant Boavista', 'count': 2},
 {'_id': '', 'count': 8},
 {'_id': 'Merkezi Konum', 'count': 2},
 {'_id': 'Cozy apartment', 'count': 2}]


In [ ]:
# Check listings with duplicated descriptions

pipeline = list(
    db.listingsAndReviews.aggregate([
        {"$group": { "_id": "$description", "count": { "$sum": 1 }} },
        { "$match": {"count": { "$gt": 1 }}}
    ])
)

pprint(pipeline)

[{'_id': 'Urban Shadow It couldn’t be better located. Walk out of the flat and '
         'you will find yourself in the heart of Tsim Sha Tsui, which is full '
         'of high end boutique shops, shopping malls, massage spas and '
         'artistic galleries. Loads of highly recommended restaurants and '
         'super cool bars are just next to the building. Feel the vibe and '
         'lively city life in Hong Kong  Urban Shadow '
         '佔據著最好的地理位置,當你走出公寓,你会发现自己位于尖沙咀的中心地带,那里有许多高端精品店,购物商场,按摩水疗馆和艺术画廊｡ '
         '强烈推荐的餐厅和超酷酒吧旁边就是大楼｡ 感受香港的氛围和生动的城市生活 In the traditional Chinese '
         'building, the simplicity and modernized design of Urban Shadow '
         'allows you to sit back and relax in the hustle and bustle yet lively '
         'city. The studio is freshly furnished. Couples, solo adventurers, '
         'and business travelers, who are tired of plain hotels, would find '
         'Urban Shadow fascinating! 在中国传统建筑中,Urban '
         'Shadow的简约和现代化设计让您在熙熙攘攘的热闹城市中放松身心

In [69]:
# Check listings with duplicated name and description

pipeline = list(
    db.listingsAndReviews.aggregate([
        {"$group": { "_id": {"description": "$description", "name": "$name"}, 
                    "count": { "$sum": 1 },
                    "documents": { "$push": "$_id" }} },

        { "$match": {"count": { "$gt": 1 }}}
    ])
)

pprint(pipeline)

[{'_id': {'description': 'Moradia concelho de  Santa Maria da feira com evento '
                         'maio teatro rua imaginárius perto das praias de  '
                         'espinho esmoriz praias Gaia zoo de lourosa e zoo de '
                         'avintes num raio de 10 kms possível de usufruir de '
                         'guia turístico com transporte Cozinha quarto de '
                         'banho terraço exterior Cozinha equipada pátio '
                         'exterior quarto de banho para duches Possibilidade '
                         'de guia para conhecer as zonas circundantes Santa '
                         'Maria da feira praias espinho Gaia Porto Praias '
                         'serra caves de Gaia zoo de avintes parque '
                         'ornotologico Carro mota táxi',
          'name': 'Quarto moradia luxo'},
  'count': 2,
  'documents': ['26556751', '26563602']},
 {'_id': {'description': '',
          'name': 'İstanbul Birden fazla bölüm

There are listings with the same name + description (one case where there is no description):
- Analyze differences to understand if it's the same property.

In [ ]:
# Visualize fields that hold different values for listings w/ same name and description

doc1 = db.listingsAndReviews.find_one({"_id": "26556751"})
doc2 = db.listingsAndReviews.find_one({"_id": "26563602"})

diffs = {}
for key in set(doc1.keys()).union(set(doc2.keys())):  
    if doc1.get(key) != doc2.get(key): 
        diffs[key] = {"doc1": doc1.get(key), "doc2": doc2.get(key)}

pprint(diffs)


{'_id': {'doc1': '26556751', 'doc2': '26563602'},
 'address': {'doc1': {'country': 'Portugal',
                      'country_code': 'PT',
                      'government_area': 'Sanguedo',
                      'location': {'coordinates': [-8.52195, 41.00466],
                                   'is_location_exact': False,
                                   'type': 'Point'},
                      'market': 'Porto',
                      'street': 'Sanguedo, Aveiro, Portugal',
                      'suburb': ''},
             'doc2': {'country': 'Portugal',
                      'country_code': 'PT',
                      'government_area': 'Sanguedo',
                      'location': {'coordinates': [-8.50928, 41.00966],
                                   'is_location_exact': False,
                                   'type': 'Point'},
                      'market': 'Porto',
                      'street': 'Sanguedo, Aveiro, Portugal',
                      'suburb': ''}},
 'ameniti

There are differences in price, however it looks like it's the same property, so we should erase the older one: 26556751.

In [73]:
# Visualize fields that hold different values for listings w/ same name and description

doc1 = db.listingsAndReviews.find_one({"_id": "28221594"})
doc2 = db.listingsAndReviews.find_one({"_id": "29224880"})
doc3 = db.listingsAndReviews.find_one({"_id": "29324861"})

diffs = {}
for key in set(doc1.keys()).union(set(doc2.keys())):  
    if doc1.get(key) != doc2.get(key): 
        diffs[key] = {"doc1": doc1.get(key), "doc2": doc2.get(key), "doc3": doc3.get(key)}

pprint(diffs)


{'_id': {'doc1': '28221594', 'doc2': '29224880', 'doc3': '29324861'},
 'accommodates': {'doc1': 1, 'doc2': 5, 'doc3': 2},
 'address': {'doc1': {'country': 'Turkey',
                      'country_code': 'TR',
                      'government_area': 'Besiktas',
                      'location': {'coordinates': [28.99908, 41.04811],
                                   'is_location_exact': False,
                                   'type': 'Point'},
                      'market': 'Istanbul',
                      'street': 'Beşiktaş, İstanbul, Turkey',
                      'suburb': 'Beşiktaş'},
             'doc2': {'country': 'Turkey',
                      'country_code': 'TR',
                      'government_area': 'Besiktas',
                      'location': {'coordinates': [29.0229, 41.05203],
                                   'is_location_exact': False,
                                   'type': 'Point'},
                      'market': 'Istanbul',
                      'stree

These listings are from different hosts and different property types, so they may be different listings even tho they have the same name.

### Reviews:

In [123]:
# Reviews with same id

pipeline = list(
    db.listingsAndReviews.aggregate([
        {"$group": { "_id": "$reviews._id", "count": { "$sum": 1 }} },
        {"$match": {"count": { "$gt": 1 }}},
    ])
)

pprint(pipeline)

[{'_id': [], 'count': 1632}]


- There are no "reviews" field with the same id.
- The output could represent documents where the reviews._id is missing or documents that don´t have any reviews at all and therefor reviews._id does not exist.


## Empty fields

In [81]:

# Check for empty fields in all documents 

cursor = db.listingsAndReviews.find()

for doc in cursor:
    empty_fields = {key: doc[key] for key in doc if doc[key] in [None, "", []]}
    
    if empty_fields:
        print(f"Document id: {doc['_id']} empty fields:")
        pprint(empty_fields)


Document id: 10009999 empty fields:
{'access': '', 'reviews': []}
Document id: 1001265 empty fields:
{'notes': ''}
Document id: 10021707 empty fields:
{'access': '',
 'house_rules': '',
 'interaction': '',
 'neighborhood_overview': '',
 'notes': '',
 'space': '',
 'transit': ''}
Document id: 10030955 empty fields:
{'access': '',
 'house_rules': '',
 'interaction': '',
 'neighborhood_overview': '',
 'notes': '',
 'reviews': [],
 'space': '',
 'transit': ''}
Document id: 1003530 empty fields:
{'interaction': '', 'summary': ''}
Document id: 10038496 empty fields:
{'notes': ''}
Document id: 10047964 empty fields:
{'access': '', 'interaction': '', 'notes': '', 'transit': ''}
Document id: 10051164 empty fields:
{'house_rules': ''}
Document id: 10057447 empty fields:
{'access': '',
 'house_rules': '',
 'interaction': '',
 'neighborhood_overview': '',
 'notes': '',
 'reviews': [],
 'transit': ''}
Document id: 10057826 empty fields:
{'interaction': '', 'notes': '', 'transit': ''}
Document id: 1

In [91]:
# Check unique fields that have at least one document where they are empty

empty_field = set()

cursor = db.listingsAndReviews.find()

for doc in cursor:
    for key in doc:
        if doc[key] in [None, "", []]: 
            empty_field.add(key)   

pprint(empty_field)


{'access',
 'description',
 'house_rules',
 'interaction',
 'name',
 'neighborhood_overview',
 'notes',
 'reviews',
 'space',
 'summary',
 'transit'}


- Several documents have empty fields, but these fields are likely not mandatory, we have no further information so we will accept it as it is.
- We can also confirm all listings are associated with a host (because the host field is not in the list), which is expected to be mandatory.

### Nested fields:

In [101]:
# Check for listings that have host_id as empty fields 

cursor = db.listingsAndReviews.find({
    "$or": [
        {"host.host_id": " "},
        {"host.host_id": None},
        {"host.host_id": []}
    ]
})

for doc in cursor:
    pprint(doc)

- All hosts have an id.

In [ ]:
# Check for listings that have reviews._id as empty fields 

cursor = db.listingsAndReviews.find({
    "$or": [
        {"reviews._id": " "},
        {"reviews._id": None},
        {"reviews._id": []}
    ]
})

for doc in cursor:
    pprint(doc)

- All reviews have an id.

In [ ]:
# Check for listings that have reviews.reviwer_id as empty fields 

cursor = list(db.listingsAndReviews.find({
    "$or": [
        {"reviews.reviwer_id": " "},
        {"reviews.reviwer_id": None},
        {"reviews.reviwer_id": []}
    ]
}))

pprint(len(cursor))

3923


There are 3923 reviews that aren't associated with an user via reviewer_id. This may indicate an inconsistency, but since we have no way to input this information, we will leave it as is.

In [142]:
# Check for listings that have reviews.listing_id as empty fields 

cursor = list(db.listingsAndReviews.find({
    "$or": [
        {"reviews.listing_id": " "},
        {"reviews.listing_id": None},
        {"reviews.listing_id": []}
    ]
}))

pprint(len(cursor))

0


- All reviews are associated with a listing via listing_id.

## Reviews' assignment

In [ ]:
# Check if the listing_id in the reviews matched the document id where the review is nested

pipeline = list(
    db.listingsAndReviews.aggregate([
        {"$unwind": "$reviews"},
        {"$match": {"$expr": { "$ne": ["$reviews.listing_id", "$_id"] }}}
    ]))

for doc in pipeline:
    pprint(doc)

- All reviews are inside the correct listing.

## Patterns:
- Computed: Add computed value average_rating for quick retrieval.
- Extended reference: When creating a new collection for "Reviews" we will keep the listing_id to link the documents across the collections.

## Apply changes to the database

TO DO:

- Create new collection for Reviews with all fields including listing_id. In the collection ListingsAndReviews change name to Listings and keep 3 most recent reviews in the documents of the listings. 
- Erase this duplicated listing: 26556751.
- Add computed value average_rating for quick retrieval, as new field for each listing document.
- Indexing.

# 1.1 Database Updates

12) Add a new property with a new host in one of the top 10 cities. The host selects the top 10 most common amenities to list.

13) Add a new review from one of our top 20 reviewers for this new property.

14) Add a new review metric called 'x_factor' with a score of 10. Show that the average score across all metrics is correctly calculated for this listing, using the previously developed query.

# 2

### Question 1

### Question 2 - Once a month, we reward hosts with recognition. Select three superhosts with at least two listings that can accommodate more than four people.

In [38]:
pipeline = list(
    db.listingsAndReviews.aggregate([
        {"$match": { 
            "host.host_is_superhost": True,
            "accommodates": { "$gt": 4 }
        }},
        {"$group": { 
            "_id": "$host.host_id", 
            "host_name": { "$first": "$host.host_name" }, 
            "host_listings_count": { "$sum": 1 }
        }},
        {"$match": { 
            "host_listings_count": { "$gte": 2 } 
        }},
        {"$sort": { "host_listings_count": -1 }}, 
        {"$limit": 3}
    ])
)

print("\nSelected Superhosts:\n")
for doc in pipeline:
    pprint(doc)



Selected Superhosts:

{'_id': '42618840',
 'host_listings_count': 3,
 'host_name': 'Great Vacation Retreats'}
{'_id': '10496350', 'host_listings_count': 3, 'host_name': 'Elite'}
{'_id': '36133410', 'host_listings_count': 3, 'host_name': 'Patty And Beckett'}


### Question 3 - The company considers investing into property to rent. Which bed type is most common in listings with a waterfront and a dishwasher in New York?

In [ ]:
pipeline = list(
    db.listingsAndReviews.aggregate([
        {"$match": { 
            "address.market": "New York",
            "amenities": {"$in": ["Dishwasher", "Waterfront"]},
        }},
        {"$group": {
            "_id": "$bed_type", 
            "count": {"$sum": 1}
        }},
        {"$sort": {"count": -1}},
        {"$limit": 1}
    ])
)

for doc in pipeline:
    pprint(doc)


{'_id': 'Real Bed', 'count': 75}


### Question 4 - We're considering hiring someone to write reviews professionally. Who wrote the longest review in New York?

In [54]:
pipeline = list(
    db.listingsAndReviews.aggregate([
        {"$match": { 
            "address.market": "New York", 
            "reviews": { "$exists": True } 
        }},
        {"$unwind": "$reviews"},
        {"$sort": { "reviews.comments": -1 }}, 
        {"$project": { 
            "reviews.reviewer_name": 1,
            "reviews.comments": 1
        }},
        {"$limit": 1}
    ])
)

print("\nPerson with the Longest Review in New York:\n")
for doc in pipeline:
    pprint(doc)



Person with the Longest Review in New York:

{'_id': '14569881',
 'reviews': {'comments': 'ﾙｰﾑｵｰﾅｰは非常に日本贔屓のｱﾒﾘｶ人で､部屋はﾃﾞｻﾞｲﾝのｱｲﾃﾞｨｱに富み､清潔で､静かで､ﾘﾗｯｸｽできます｡\n'
                         'ﾙｰﾑﾒｲﾄも皆優しくて親切です｡\n'
                         '泊まった部屋は､大きなｽﾀｼﾞｵの中に手作りで創られた2階建ての小屋の1階でした｡工夫と創造性に溢れた､素敵な空間です｡wifiもあります｡\n'
                         'ただ､もし一般的ﾋﾞｼﾞﾈｽﾕｰｽで考えるなら､この部屋に限らず､一般的なｼｪｱﾙｰﾑは避けたほうがいいでしょう｡日本人はなんだかんだで､仕事に集中したい時は完全に遮蔽されたﾌﾟﾗｲﾍﾞｰﾄ空間を好みます(僕もそう)が､ｼｪｱﾙｰﾑでそれは無理です｡静かとは言え､3人が共同生活をしているので､会話は発生するし､料理や掃除も始まるし､風呂の順番だってﾏﾅｰを守らねばなりません｡あくまでくつろぐときに部屋を利用し､仕事はｶﾌｪなどでやるのが無難です｡\n'
                         'ちなみに､ｵｰﾅｰは日本語を話せますが､少しだけです｡基本的に英語でｺﾐｭﾆｹｰｼｮﾝする必要があります｡\n'
                         'ちなみにｱﾒﾆﾃｨはﾊﾞｽﾀｵﾙ1枚なので､歯ﾌﾞﾗｼや石鹸など身の回りのものは自分で持っていくか近所のﾃﾞﾘなどで買いましょう｡\n'
                         'よい旅を｡',
             'reviewer_name': 'Kazu'}}


### Question 5 - To assess the security of different areas, what is the biggest and smallest (price-security deposit) difference per number of visitors at a property?

In [ ]:
pipeline = list(
    db.listingsAndReviews.aggregate([
        {"$match": {
            "price": {"$exists": True},
            "security_deposit": {"$exists": True}  
        }},

        {"$project": {
            "difference": {"$subtract": ["$price", "$security_deposit"]} 
        }},
        {"$group": {
            "_id": None, 
            "max_difference": {"$max": "$difference"}, 
            "min_difference": {"$min": "$difference"}
        }}
    ])
)

print("\nBiggest and Smallest Price-Security Deposit Differences:\n")
for doc in pipeline:
    pprint(doc)




Biggest and Smallest Price-Security Deposit Differences:

{'_id': None,
 'max_difference': Decimal128('8501.00'),
 'min_difference': Decimal128('-38459.00')}


In [ ]:
#mesma coisa mas a retornar os ID's 
pipeline = list(
    db.listingsAndReviews.aggregate([
        {"$match": {
            "price": {"$exists": True},  
            "security_deposit": {"$exists": True}  
        }},
        {"$project": {
            "_id": 1,  
            "price": {"$toDouble": "$price"},  
            "security_deposit": {"$toDouble": "$security_deposit"},  
            "difference": {"$subtract": [{"$toDouble": "$price"}, {"$toDouble": "$security_deposit"}]}  
        }},
        {"$sort": {"difference": 1}}, 
        {"$group": {
            "_id": None,  
            "max_difference": {"$last": "$difference"},  
            "max_id": {"$last": "$_id"},
            "min_difference": {"$first": "$difference"},  
            "min_id": {"$first": "$_id"} 
        }}
    ])
)

print("\nBiggest and Smallest Price-Security Deposit Differences:\n")
for doc in pipeline:
    pprint(doc)



Biggest and Smallest Price-Security Deposit Differences:

{'_id': None,
 'max_difference': 8501.0,
 'max_id': '22200454',
 'min_difference': -38459.0,
 'min_id': '2688143'}


In [ ]:
#to check price-security_deposit difference
document = db.listingsAndReviews.find({"_id": "2688143"})

print("\nDocument for _id = 22200454:\n")
for doc in document:
    pprint(doc)



Document for _id = 22200454:

{'_id': '2688143',
 'access': '',
 'accommodates': 2,
 'address': {'country': 'Hong Kong',
             'country_code': 'HK',
             'government_area': 'Eastern',
             'location': {'coordinates': [114.19337, 22.28558],
                          'is_location_exact': True,
                          'type': 'Point'},
             'market': 'Hong Kong',
             'street': 'Hong Kong, Hong Kong Island, Hong Kong',
             'suburb': 'Eastern District'},
 'amenities': ['TV',
               'Internet',
               'Wifi',
               'Air conditioning',
               'Kitchen',
               'Paid parking off premises',
               'Elevator',
               'Washer',
               'Dryer',
               'Hangers',
               'Hair dryer',
               'Iron',
               'Laptop friendly workspace',
               'Room-darkening shades',
               'Hot water',
               'Bed linens',
               'Etherne

### Question 6

### Question 7

### Question 8

### Question 9

### Question 10

### Question 11

### Question 12

### Question 13

### Question 14